File: problem-114/solution.go
package problem114

import (
	"fmt"
	"regexp"
	"strings"
)

func reverse(s string, delimiters string) string {
	// Get words
	delimiterRegex := regexp.MustCompile(fmt.Sprintf("[%s]+", delimiters))
	words := delimiterRegex.Split(s, -1)

	// Get delimiters
	nonDelimiterRegex := regexp.MustCompile(fmt.Sprintf("[^(%s)]+", delimiters))
	nonWords := nonDelimiterRegex.Split(s, -1)

	var result strings.Builder

	// Check if last word is empty (when there is a delimiter in the end)
	wordPos := len(words) - 1
	if words[wordPos] == "" {
		wordPos--
	}

	// Combine reversed words and the delimiters
	for _, delimiter := range nonWords {
		result.WriteString(delimiter)
		if wordPos < 0 {
			continue
		}
		result.WriteString(words[wordPos])
		wordPos--
	}

	return result.String()
}

File: problem-114/solution.py
from typing import List
from re import split
from unittest import TestCase


def reverse(s: str, delimiters: List[str]) -> str:
    words = split('[' + delimiters + ']+', s)
    not_words = split('[^(' + delimiters + ')]+', s)
    print(s, words, not_words)
    if len(words) > 0 and words[-1] == "":
        words = words[:-1]

    words_iterator = reversed(words)

    result: List[str] = []

    for delimiter in not_words:
        result.append(delimiter)
        try:
            result.append(next(words_iterator))
        except StopIteration:
            pass

    return ''.join(result)


class Test(TestCase):
    def test(self):
        test_cases = [
            ("hello/world:here", ":/", "here/world:hello"),
            ("hello/world:here/", ":/", "here/world:hello/"),
            ("hello//world:here", ":/", "here//world:hello"),
        ]

        for test_case in test_cases:
            self.assertEqual(test_case[2], reverse(test_case[0], test_case[1]))
